### ISSUE TITLE
preferredTextLanguage does not affect selection flags for text tracks

### ISSUE THREAD

**Issue by petersamokhin** [google/ExoPlayer#9953]
## TL;DR
Now there is no way to get the actual text track which is not playing when the preferred language set.
Instead, the default track is marked as selected.
ExoPlayer version: `2.16.0`

### Media
Have an HLS video with more than one language. 
One of the languages is marked as default in the manifest.
Example: https://bitdash-a.akamaihd.net/content/sintel/hls/playlist.m3u8

### Setup preferred language
Set the preferred language to non-default:
```kotlin
val selector = DefaultTrackSelector()

// ...

fun setPreferredLanguage(lang: String) {
  selector
    .setParameters(
      selector.buildUponParameters()
        .setPreferredTextLanguages(lang)
      )
}
```

The `setPreferredTextLanguageAndRoleFlagsToCaptioningManagerSettings` is not relevant here as the accessibility feature may not be enabled but we still want to use the e.g. device language for displaying subtitles.

### Get the text tracks
```kotlin
private val MappingTrackSelector.MappedTrackInfo.textRendererIndex: Int?
  get() = (0 until rendererCount)
    .find { rendererIndex -> getRendererType(rendererIndex) == C.TRACK_TYPE_TEXT }

// ...

val info = currentMappedTrackInfo!!
val rendererIndex = info.textRendererIndex!!
val trackGroups = info.getTrackGroups(rendererIndex)
``` 
`TrackGroup` -> `Format` -> `selectionFlags` is `1` for the default language and `0` for the others (including the actual track which the user sees). So here the information is wrong.

### Selection override
```kotlin
val override = selector.parameters.getSelectionOverride(rendererIndex, trackGroups)
// is null
```
The preferred language might be treated as the selection override, but no information available regarding the actual selected track.

**Comment by ojw28** [comment]
> TrackGroup -> Format -> selectionFlags is 1 for the default language and 0 for the others (including the actual track which the user sees). So here the information is wrong.

The information in the format indicates what the media specifies, not what the application or user has selected during playback.

> The preferred language might be treated as the selection override, but no information available regarding the actual selected track.

If you take a look at the logging output by our demo app, you'll see an `[X]` marking selected tracks, and that this is correctly updated to indicate the selected track. So the information is definitely available.

I think you should take a look at the information you can query through `Player.getCurrentTracksInfo()`, and in particular `TrackGroupInfo.isTrackSelected(int trackIndex)`. If you're using an earlier version of ExoPlayer that doesn't have this API, you can look at what the `EventLogger` class does in that release to identify which tracks are selected.

**Comment by petersamokhin** [comment]
@ojw28 thanks for your reply. I see that we can retrieve this info asynchronously but we need then to have a reference to the player, just having the track selector won't be enough. 
Maybe it will make sense to treat the predefined selection based on the preferred languages as the selection override (which seems logical) as the track selector seems to be the entity responsible for the tracks selection (😮) but is not enough to gather all the info required for e.g. asynchronous display of the tracks list (also showing which one is selected) to the user.
It looked to me as a delegate just for such purposes but it isn't.
You can treat it as a suggestion but this question can be closed, thanks!

**Comment by ojw28** [comment]
> I see that we can retrieve this info asynchronously but we need then to have a reference to the player, just having the track selector won't be enough.

The story with ExoPlayer's track selection APIs has been a little confusing, but the direction of travel is that you should be able to query available tracks and select the ones you want via the `Player` API, unless you have some particularly unusual/niche use case. The `TrackSelector` is intended to be an internal component used by the player (similar to `Renderer` instances, which you'd not generally try to query directly). We're not quite fully there yet, but that's where we're headed.

Note that recent releases have `Player.getTrackSelectionParameters` and `Player.setTrackSelectionParameters` to help with this. We still need to move `TrackSelectionDialog` over to use the `Player` API. Once we've done that the demo app will stop holding a reference to a `TrackSelector`, and the UI module will stop depending on the `core` module that contains the `ExoPlayer` implementation of `Player`, and things should be a lot clearer.
